### PR TITLE
Clean up the Timer CPU time calculations.

### PR DESCRIPTION
**Pull request by drwells**
More progress towards #4867: this collects all of the (platform-dependent) resource usage code in one place.

### CODE REVIEW

**Comment by drwells** [comment]
A comment for the reviewer: all of these `std::chrono` calculations are currently over-engineered (I convert a double to a `chrono` type, do a calculation, and then convert back). The next patch will convert all time measurement members members of `Timer` from `double to either chrono durations or chrono time points and these statements will make a lot more sense in that context.

File: source/base/timer.cc [hunk 1]
```diff
@@ -37,6 +38,83 @@
...
+        return T(std::lround(T::period::den*(time/T::period::num)));
+      }
+
+      /**
+       * Convert a given duration into a double precision number with units of
+       * seconds.
+       */
+      template <typename Rep, typename Period>
+      double
+      duration_to_double(const std::chrono::duration<Rep, Period> duration)
```
Comment (masterleinad) [inline source/base/timer.cc]: and here `duration_to_seconds`.

File: source/base/timer.cc [hunk 2]
```diff
@@ -37,6 +38,83 @@
...
+      template <typename Rep, typename Period>
+      struct is_duration<std::chrono::duration<Rep, Period>> : std::true_type {};
+
+      /**
+       * Convert a double precision number with units of seconds into a
+       * specified duration type T.
+       */
+      template <typename T>
+      T
+      double_to_duration(const double time)
```
Comment (masterleinad) [inline source/base/timer.cc]: A better name might be `seconds_to_duration`.

File: source/base/timer.cc [hunk 3]
```diff
@@ -179,24 +242,11 @@ double Timer::cpu_time() const
 {
   if (running)
     {
-#if defined(DEAL_II_HAVE_SYS_TIME_H) && defined(DEAL_II_HAVE_SYS_RESOURCE_H)
-      rusage usage;
-      getrusage (RUSAGE_SELF, &usage);
-      const double dtime =  usage.ru_utime.tv_sec + 1.e-6 * usage.ru_utime.tv_usec;
-      const double running_time = dtime - current_lap_starting_cpu_time + accumulated_cpu_time;
-
-      // in case of MPI, need to get the time passed by summing the time over
```
Comment (masterleinad) [inline source/base/timer.cc]: Is there a reason to delete this comment?

File: source/base/timer.cc [hunk 4]
```diff
@@ -13,16 +13,17 @@
 //
 // ---------------------------------------------------------------------
 
-#include <deal.II/base/timer.h>
 #include <deal.II/base/exceptions.h>
+#include <deal.II/base/timer.h>
```
Comment (masterleinad) [inline source/base/timer.cc]: Looks like sorting alphabetically. Why is this out of order?

File: source/base/timer.cc [hunk 5]
```diff
@@ -13,16 +13,17 @@
...
-#include <sstream>
+#include <deal.II/base/utilities.h>
+
+#include <algorithm>
+#include <chrono>
 #include <iostream>
 #include <iomanip>
-#include <algorithm>
-#include <stddef.h>
+#include <sstream>
```
Comment (masterleinad) [inline source/base/timer.cc]: ... and <iomanip> and <iostream>. Or did I miss the point of moving these headers totally?

File: source/base/timer.cc (hunk 4 above)
Comment (drwells) [inline source/base/timer.cc]: Because I forgot to fix it.

File: source/base/timer.cc (hunk 5 above)
Comment (drwells) [inline source/base/timer.cc]: `<algorithm>` is now higher up; we never use anything in `stddef.h` so I removed it.

File: source/base/timer.cc [hunk 6]
```diff
@@ -37,6 +38,83 @@
...
+{
+  namespace Timer
+  {
+    namespace
+    {
+      /**
+       * Type trait for checking whether or not something is a duration.
+       */
+      template <typename T>
+      struct is_duration : std::false_type {};
```
Comment (drwells) [inline source/base/timer.cc]: None; this is just a check on my own work to make sure I never pass in a non-duration.

File: source/base/timer.cc [hunk 7]
```diff
@@ -37,6 +38,83 @@
...
+
+      /**
+       * Specialization for durations.
+       */
+      template <typename Rep, typename Period>
+      struct is_duration<std::chrono::duration<Rep, Period>> : std::true_type {};
```
Comment (drwells) [inline source/base/timer.cc]: All duration types inherit from some instantiation of `std::ratio<M,N>` (which gives the `period`; the underlying integral type is `Rep`). I am not sure about the name of the private value member accessed by `count()`; did you mean something else by '`duration` member'?

File: source/base/timer.cc (hunk 1 above)
Comment (drwells) [inline source/base/timer.cc]: (same)

File: source/base/timer.cc (hunk 3 above)
Comment (drwells) [inline source/base/timer.cc]: Since this is always done I think a better place to describe what is going on is the documentation (which I will update).

**Comment by drwells** [comment]
I have made the requested changes (and clarified the MPI comment in the method documentation, rather than in the source code).

**Comment by drwells** [comment]
There are a few other small changes I would like to make (I might change around some of the internal functions) so don't merge this yet.

File: source/base/timer.cc (hunk 6 above)
Comment (masterleinad) [inline source/base/timer.cc]: Can you then just say `std::chrono::duration` in the doxygen comment as well?

File: source/base/timer.cc (hunk 7 above)
Comment (masterleinad) [inline source/base/timer.cc]: If you only intent to check that the given type is in fact a `std::chrono::duration` type, I agree with this type trait.

File: source/base/timer.cc (hunk 6 above)
Comment (drwells) [inline source/base/timer.cc]: I'll add that.

**Comment by drwells** [comment]
I went back and changed the names to `to_seconds` and `from_seconds`; including the type name in the function name is redundant (and much more verbose).

**Comment by drwells** [comment]
Clarification: this is ready for re-review. I only further shortened the names of some functions, so this should be easy to check.

**Comment by masterleinad** [comment]
Thanks @drwells !

**Pull request merged**